## Slide 40 - Gramática

### Definição Informal

**Gramáticas** - Também conhecidas como **dispositivos generativos**, **dispositivos de síntese** ou **dispositivos de geração de cadeias**, as gramáticas constituem sistemas formais baseados em regras de substituição, através dos quais é possível sintetizar, de forma exaustiva, o conjunto das cadeias que compõem uma determinada linguagem.


Por exemplo, em português, tem-se, dentre outras, a seguinte regra de formação de sentenças: "uma sentença pode consistir de uma frase nominal seguida de um predicado". Concisamente, tem-se:

- `<sentenca> -> <frase_nominal> + <predicado>`

A definição acima pode ser mais precisa, fazendo-se:

- `<frase_nominal> -> <artigo> <substantivo>,`
- `<predicado> -> <verbo>`

Ao associar palavras às classes sintáticas, podem-se formar sentenças:

- `<artigo>`: a, o, um, uma
- `<substantivo>`: menino, cachorro
- `<verbo>`: anda, corre

Para exemplificar a gramática em código, podemos criar uma simples implementação em Python que gera sentenças com base nas regras definidas:

In [ ]:
import random

artigos = ['a', 'o', 'um', 'uma']
substantivos = ['menino', 'cachorro']
verbos = ['anda', 'corre']

def gerar_sentenca():
    artigo = random.choice(artigos)
    substantivo = random.choice(substantivos)
    verbo = random.choice(verbos)
    return f"{artigo} {substantivo} {verbo}"

# Exemplo de geração de sentença
print(gerar_sentenca())

a cachorro anda


### Definição Formal

Formalmente, uma gramática G pode ser definida como sendo uma quádrupla:

G = (V, $\Sigma$, P, S)

onde:

- \(V\): é o **vocabulário** da gramática; corresponde a um conjunto finito e não vazio de símbolos.
    - Pense no vocabulário como todas as palavras que você pode usar em uma língua. Em gramática, são os símbolos com os quais você trabalha, incluindo tanto os terminais quanto os não terminais.
- \($\Sigma$\): é o conjunto finito e não vazio dos **símbolos terminais** da gramática. É o alfabeto.
    - Estes são os elementos básicos, como as letras do alfabeto, que você não pode mais dividir em partes menores.
- \(P\): é o conjunto finito e não vazio de **produções** ou **regras de substituição** da gramática.
    - São as regras que dizem como você pode combinar ou transformar os símbolos para formar sentenças válidas na linguagem.
- \(S\): é a **raiz** da gramática, \(S $\in$ V\).
    - Este é o ponto de partida da gramática, o símbolo a partir do qual todas as sentenças são derivadas.
- \(N = V - $\Sigma$\): é o conjunto de **símbolos não terminais** da gramática. São as classes sintáticas.
    - Estes são os símbolos que podem ser substituídos ou expandidos em outros símbolos, servindo como intermediários na construção de sentenças.

Podemos ilustrar esses conceitos com um código simples que define uma gramática básica:

In [11]:
# Definição da gramática G
V = {'S', 'A', 'B', 'a', 'b'}  # Vocabulário
Sigma = {'a', 'b'}             # Símbolos terminais
P = {                          # Produções
    'S': ['A B'],
    'A': ['a'],
    'B': ['b']
}
S = 'S'                        # Raiz

# Símbolos não terminais
N = V - Sigma

print("Vocabulário:", V)
print("Símbolos terminais:", Sigma)
print("Produções:", P)
print("Raiz:", S)
print("Símbolos não terminais:", N)

Vocabulário: {'b', 'S', 'A', 'a', 'B'}
Símbolos terminais: {'b', 'a'}
Produções: {'S': ['A B'], 'A': ['a'], 'B': ['b']}
Raiz: S
Símbolos não terminais: {'B', 'S', 'A'}


### Forma das Produções Gramaticais

O conjunto $P$ de produções gramaticais obedece à forma geral:

- $\alpha$ $\rightarrow$ $\beta$, com $\alpha$ $\in$ $V^*NV^*$ e $\beta$ $\in$ $V^*$

onde:

- $\alpha$ é uma cadeia constituída de quaisquer combinações de símbolos de $V$, contendo pelo menos um símbolo não-terminal.
- $\beta$ é uma cadeia qualquer, eventualmente vazia, de elementos de $V$.

Portanto, o conjunto $P$ pode ser expresso como uma relação:

- $P$ = { ($\alpha$, $\beta$) | ($\alpha$, $\beta$) $\in$ $V^*NV^*$ $\times$ $V^*$ })
    - $P$: As regras que definem como os símbolos da gramática podem ser combinados ou transformados. Elas são essenciais para construir sentenças na linguagem que a gramática descreve.
    - $\alpha$ $\rightarrow$ $\beta$: Esta notação descreve uma regra de produção onde α pode ser substituído por β.
    - α deve ser uma sequência de símbolos que inclui pelo menos um símbolo não terminal (indicado por N). $V^∗$ significa qualquer sequência de símbolos do vocabulário, incluindo a sequência vazia.
    - $\beta$ $\in$ $V^*$: pode ser qualquer sequência de símbolos do vocabulário, incluindo a possibilidade de ser uma sequência vazia (ou seja, não ter nenhum símbolo).

#### Exemplo 

Considere uma gramática simples com as seguintes características:

- Vocabulário V = \{$\text{"A"}$, $\text{"B"}$, $\text{"a"}$, $\text{"b"}$\}, onde "A" e "B" são símbolos não terminais e "a" e "b" são símbolos terminais.
- Símbolo inicial $S$ é "A".
- As produções $P$ são definidas como:
  - $A \rightarrow aB$: Aqui, "A" pode ser substituído por "aB".
  - $B \rightarrow b$ : Aqui, "B" pode ser substituído por "b".

Seguindo estas regras, podemos formar a sentença "ab" ao aplicar as produções:

1. Substituímos "A" por "aB" (de acordo com a primeira regra de produção).
2. Em seguida, substituímos "B" por "b" (conforme a segunda regra), resultando na sentença final "ab".

Vamos representar essa gramática em Python e gerar a sentença conforme as regras de produção, onde:

- $V$: Conjunto de símbolos, tanto terminais ($'a'$, $'b'$) quanto não terminais ($'A'$, $'B'$). O vocabulário da gramática.
- $P$: Dicionário representando as regras de produção. Cada chave é um símbolo não terminal, e o valor é uma lista de substituições possíveis. Por exemplo, $'A': ['aB']$ indica que $A$ pode ser substituído por $aB$.
- $S$: O símbolo inicial da gramática. A geração de sentenças começa a partir deste símbolo.

In [ ]:
# Definição da gramática
V = {'A', 'B', 'a', 'b'}  # Vocabulário
P = {                     # Produções
    'A': ['aB'],
    'B': ['b']
}
S = 'A'                   # Símbolo inicial

# Função para aplicar as produções e gerar a sentença
def gerar_sentenca(S, P):
    sentenca = S
    # sentenca = aB
    # symbol = a
    while any(symbol in P for symbol in sentenca):
        
        for symbol in sentenca:
            # symbol = B
            # P = {'A': ['aB'], 'B': ['b']}
            if symbol in P:
                # Substitui o símbolo não terminal pela sua produção
                sentenca = sentenca.replace(symbol, P[symbol][0], 1)
                break
    return sentenca

# Gerar e imprimir a sentença
sentenca_gerada = gerar_sentenca(S, P)
print("Sentença gerada:", sentenca_gerada)

Sentença gerada: ab


- `gerar_sentenca(S, P)`: Função que gera sentenças a partir do símbolo inicial $S$ usando as regras de produção $P$.
- Inicia com sentenca sendo o símbolo inicial $S$.
- Continua substituindo os símbolos não terminais em sentenca pelas suas substituições em $P$ até que não haja mais símbolos não terminais.
- `while any(symbol in P for symbol in sentenca)`: Loop que continua enquanto houver símbolos não terminais em sentenca que podem ser substituídos.
- Dentro do loop, `for symbol in sentenca` percorre cada símbolo de sentenca.
- Se o símbolo atual está em $P$, substitui a primeira ocorrência desse símbolo na sentença por sua produção em $P$.
- A substituição é feita uma vez por iteração do loop `while` para garantir o processamento adequado de todos os símbolos não terminais.
Retorna a sentenca final gerada.

### Forma Sentencial

**Forma sentencial**: é qualquer cadeia obtida pela aplicação recorrente das regras de substituição:

1. A raiz $S$ da gramática é por definição uma forma sentencial.
2. Seja $\alpha\rho\beta$ uma forma sentencial, com $\alpha$ e $\beta$ cadeias quaisquer de terminais e/ou não-terminais, e seja $\rho \to \gamma$ uma produção da gramática. Então, dessa regra àquela forma sentencial, substituindo a ocorrência de $\rho$ por $\gamma$, produz uma nova forma sentencial $\alpha\gamma\beta$.

Esta forma de substituição é chamada de **derivação direta** e é denotada por:

$\alpha\rho\beta \underset{G}{\implies} \alpha\gamma\beta$

Em outras palavras:
- Uma forma sentencial em uma gramática é uma sequência de símbolos que pode ser gerada a partir do símbolo inicial (raiz) pela aplicação sucessiva das regras de produção da gramática. 
- A definição de forma sentencial e derivação direta aqui nos diz que, se você tem uma cadeia de símbolos (incluindo terminais e não terminais) e uma regra de produção aplicável, você pode substituir o não terminal na cadeia pela sua produção para obter uma nova cadeia, representando um passo na geração de sentenças na linguagem definida pela gramática.

Como exemplo de especificação de gramáticas, tem-se $G_1 = (V_1, \Sigma_1, P_1, S)$, com:

- $V_1 = \{ 0,1,2,3,S,A \}$
- $\Sigma_1 = \{ 0, 1, 2, 3 \}$
- $N_1 = \{ S, A \}$
- $P_1 = \{ S \to 0S33, S \to A, A \to 12, A \to \epsilon \}$
onde:
1. $V_1$ 
    - é o conjunto de todos os símbolos (terminais e não-terminais) na gramática. Esses símbolos são os elementos básicos usados para construir as sentenças na linguagem definida pela gramática.
1. $\Sigma_1$ 
    - é o conjunto de símbolos terminais. Estes são os elementos que não podem ser substituídos por outras regras na gramática e aparecem na forma final das sentenças.
1. $N_1$ 
    - representa os símbolos não-terminais, que são os símbolos que podem ser substituídos conforme as regras de produção. Eles servem como placeholders ou variáveis que são refinados em símbolos mais específicos (terminais ou não terminais) através do processo de derivação.
1. $P_1$
    - são as regras de produção que definem como os símbolos não-terminais podem ser transformados ou substituídos para formar cadeias na linguagem. Cada regra de produção indica como um símbolo não terminal pode ser expandido ou convertido em uma sequência de símbolos terminais e/ou não terminais.


### Derivação


- **Derivação**: É a sequência de zero ou mais derivações diretas $\alpha \implies \beta \ldots \implies \mu$. É denotada por $\alpha \overset{*}{\implies} \mu$.
- **Derivação não-trivial**: É aquela em que ocorre a aplicação de pelo menos uma produção. É denotada por $\alpha \overset{+}{\implies} \mu$.
- **Sentença**: Se, pela aplicação de uma derivação não-trivial à raiz `S` de uma gramática, for possível obter uma cadeia `w` formada exclusivamente de símbolos terminais, diz-se que `w`, além de ser uma forma sentencial, é também uma sentença, e denota-se a sua derivação por $S \overset{+}{\implies} w$.

**Exemplo**

Considere uma gramática $G_1$, com as seguintes regras de produções:
- $S \to 0S33$
- $S \to A$

- $S$ é uma forma sentencial.
- $0S33$ é uma forma sentencial, pois $S \implies 0S33$.
- $S \implies 0S33$ é uma derivação direta.
- 0`0S33`33 e 00`A`3333 são formas sentenciais, pois $0S33 \implies 00S3333 \implies 00A3333$ através das produções $S \to 0S33$ e $S \to A$, aplicadas nesta ordem.

No exemplo com $G1$, começamos com $S$ e aplicamos as regras de produção para obter cadeias mais longas. Cada aplicação de uma regra é uma derivação direta, e juntas formam uma derivação que leva a uma forma sentencial ou uma sentença completa se a cadeia final contiver apenas símbolos terminais. Portanto, 0`S`33, 0`0S33`33 e 0`0A33`33 são todas formas sentenciais derivadas de $S$ usando as regras da gramática $G1$.


#### Explicação Detalhada para Iniciantes

- Derivação: Imagine que você está montando um quebra-cabeça, e a cada passo, você encaixa uma peça. Cada peça que você coloca é como uma etapa na derivação: você está construindo sua sentença, peça por peça (ou símbolo por símbolo).
- Derivação não-trivial: Uma derivação não-trivial significa que você realmente colocou pelo menos uma peça no quebra-cabeça. Em termos de gramática, significa que pelo menos uma regra de produção foi aplicada para transformar um símbolo não terminal em outro conjunto de símbolos.
- Sentença: Quando todas as peças do quebra-cabeça estão encaixadas, e você vê a imagem completa, você tem uma sentença completa na gramática. Isso significa que, começando pelo símbolo inicial $S$, você aplicou regras de produção suficientes para chegar a uma sequência de símbolos terminais (as peças finais do quebra-cabeça).

#### Questão 1

Escreva uma função em Python chamada `gerar_forma_sentencial` que, dada uma gramática e um símbolo inicial, gera uma nova forma sentencial realizando apenas uma derivação direta a partir do símbolo inicial.

A gramática é definida pelo seguinte conjunto de regras e símbolos:

- **Símbolos Não Terminais (N)**: `S`, `A`
  - Estes são os símbolos que podem ser substituídos conforme as regras de produção da gramática.

- **Símbolos Terminais (Σ)**: `0`, `1`, `3`
  - Estes são os símbolos que aparecem na forma final das sentenças e não são substituídos por outras regras.

- **Produções (P)**:
  - `S -> 0S33`: Esta regra indica que o símbolo não terminal `S` pode ser substituído por `0S33`.
  - `S -> A`: Esta regra indica que o símbolo não terminal `S` também pode ser substituído por `A`.
  - `A -> 1`: Esta regra indica que o símbolo não terminal `A` pode ser substituído por `1`.

- **Símbolo Inicial**: `S`
  - `S` é o ponto de partida para a geração de sentenças na linguagem definida pela gramática.

ou, 

$G = (V, \Sigma, P, S)$, com:

- $V = \{ 0, 1, 3, S, A \}$
- $\Sigma = \{ 0, 1, 3 \}$
- $N = \{ S, A \}$
- $P = \{ S \to 0S33, S \to A, A \to 1 \}$


Este conjunto de regras forma a gramática. Em python, realizar derivações para somente o 'S' e sua primeira possibilidade, que pode ser representada em Python como:

```python
P = {'S': ['0S33', 'A'], 'A': ['1']}
simbolo_inicial = 'S'


In [6]:
def gerar_forma_sentencial(P, simbolo_inicial):
    # Verifica se o símbolo inicial tem produções na gramática
    #S
    if simbolo_inicial in P:
        # Escolhe a primeira produção disponível para o símbolo inicial
        # P[S][0] = 0S33
        producao = P[simbolo_inicial][0]
        # Gera a forma sentencial substituindo o símbolo inicial pela produção
        forma_sentencial = simbolo_inicial.replace(simbolo_inicial, producao)
        return forma_sentencial
    return simbolo_inicial  # Retorna o próprio símbolo se não houver produções

# Exemplo de uso
P = {'S': ['0S33', 'A'], 'A': ['1']}
simbolo_inicial = 'S'
forma_sentencial = gerar_forma_sentencial(P, simbolo_inicial)
print("Forma sentencial gerada:", forma_sentencial)


Forma sentencial gerada: 0S33


- A função gerar_forma_sentencial aceita uma gramática (gramatica) e um símbolo inicial (simbolo_inicial).
- Primeiro, verifica se o símbolo inicial tem regras de produção associadas na gramática.
- Se houver produções disponíveis, a função seleciona a primeira produção da lista para esse símbolo.
- Então, substitui o símbolo inicial na forma sentencial pela produção escolhida, criando assim uma nova forma sentencial.
- Se o símbolo inicial não tiver produções na gramática, a função simplesmente retorna o próprio símbolo inicial, indicando que nenhuma derivação foi possível.
- O resultado é a nova forma sentencial após aplicar uma única derivação direta a partir do símbolo inicial.

#### Questão 1

Escreva uma função em Python chamada gerar_derivações que recebe a gramática anterior, um símbolo inicial, e um número máximo de derivações. A função deve gerar todas as formas sentenciais possíveis para a primeira substituição (S e sua primeira substituição somente) dentro do limite de derivações. Exemplo: S no nível 0, 0S33, no nível 1, 00S3333 no nível 2, 000S333333 no nivel 3 e assim por diante.

In [ ]:
def gerar_forma_sentencial(P, forma_sentencial):
    nova_forma = forma_sentencial
    # S
    # 0S33
    for simbolo in forma_sentencial:
        if simbolo in P:
            producao = P[simbolo][0]
            nova_forma = nova_forma.replace(simbolo, producao, 1)
            # 0S33
            # 00S3333
            break  # Apenas uma substituição por derivação
    return nova_forma

def gerar_derivacoes(P, simbolo_inicial, n):
    forma_atual = simbolo_inicial
    for _ in range(n):
        # S
        # 0S33
        # 0 0S33 33
        # 00 0S33 3333
        forma_atual = gerar_forma_sentencial(P, forma_atual)
    return forma_atual

# Exemplo de uso
P = {'S': ['0S33', 'A'], 'A': ['1']}
simbolo_inicial = 'S'
num_derivacoes = 2
forma_sentencial = gerar_derivacoes(P, simbolo_inicial, num_derivacoes)
print("Forma sentencial após", num_derivacoes, "derivações:", forma_sentencial)


Forma sentencial após 2 derivações: 00S3333


- `gerar_forma_sentencial`: Esta função gera a próxima forma sentencial pela aplicação da primeira produção encontrada para o símbolo inicial.
- `gerar_derivacoes`: Esta função controla o número de derivações a serem realizadas. Ela inicia com o simbolo_inicial e aplica a função gerar_forma_sentencial repetidamente, conforme o número de derivações (n) especificado.
- A cada iteração do loop, forma_atual é atualizada para a nova forma sentencial gerada pela função gerar_forma_sentencial.
- Após completar o número de derivações desejado, a última forma sentencial é retornada.

#### Questão 2

Escreva uma função em Python chamada gerar_formas_sentenciais que, dada uma gramática simples e um símbolo inicial, gera `todas` as derivações `somente para um determinado nível de derivações`, ou seja, derivações para todos os símbolos deriváveis no nível N. Por exemplo, dada as regras ${'S': ['0S33', 'A'], 'A': ['1']}$ e o símbolo inicial $S$, a função deve ser capaz de gerar formas sentenciais como {'S'} para o nível 0, {'0S33', 'A'} para o nível 1, {'0A33', '00S3333', '1'} para o nível 2, e assim por diante.

In [ ]:
def gerar_forma_sentencial(P, forma_sentencial):
    nova_forma = []
    # forma_sentencial = S
    for simbolo in forma_sentencial:
        # simbolo = S
        if simbolo in P:
            # producao = 0S33
            for producao in P[simbolo]:
                nova_forma.append(forma_sentencial.replace(simbolo, producao, 1))
                # nova_forma.append('0S33')
                # nova_forma.append('A')
                ## 1
            break  # Para considerar a primeira substituição possível para o símbolo
    return nova_forma if nova_forma else [forma_sentencial]

def gerar_derivacoes(P, simbolo_inicial, n):
    formas_sentenciais = [simbolo_inicial]
    for _ in range(n):
        novas_formas = []
        # Formas sententenciais = ['S']
        for forma in formas_sentenciais:
            # forma = S
            # novas_formas = ['']
            novas_formas.extend(gerar_forma_sentencial(P, forma))
            # novas_formas = ['0S33', `A`]
        formas_sentenciais = novas_formas
        # formas_sentenciais = ['S']
    return set(formas_sentenciais)

# Exemplo de uso
P = {'S': ['0S33', 'A'], 'A': ['1']}
simbolo_inicial = 'S'
max_derivacoes = 1
formas_sentenciais = gerar_derivacoes(P, simbolo_inicial, max_derivacoes)
print("Formas sentenciais geradas:", formas_sentenciais)


Formas sentenciais geradas: {'0S33', 'A'}


#### Questão 3

Escreva uma função em Python chamada gerar_formas_sentenciais que, dada uma gramática simples e um símbolo inicial, gera as `todas` as derivações `para todos os níveis de derivações`, ou seja, derivações para todos os símbolos deriváveis até nível N. Por exemplo, dada as regras ${'S': ['0S33', 'A'], 'A': ['1']}$ e o símbolo inicial $S$, a função deve ser capaz de gerar formas sentenciais como {'S'} para o nível 0, {'0S33', 'S', 'A'} para o nível 1, {'0A33', '0S33', 'A', '1', 'S', '00S3333'} para o nível 2, e assim por diante.
Dica: Use o código da questão anterior e modifique uma única linha!

In [14]:
def gerar_forma_sentencial(gramatica, forma_sentencial):
    nova_forma = []
    for simbolo in forma_sentencial:
        if simbolo in gramatica:
            for producao in gramatica[simbolo]:
                nova_forma.append(forma_sentencial.replace(simbolo, producao, 1))
            break  # Para considerar a primeira substituição possível para o símbolo
    return nova_forma if nova_forma else [forma_sentencial]

def gerar_derivacoes(gramatica, simbolo_inicial, n):
    formas_sentenciais = [simbolo_inicial]
    for _ in range(n):
        novas_formas = []
        for forma in formas_sentenciais:
            novas_formas.extend(gerar_forma_sentencial(gramatica, forma))
        # Linha modificada
        # formas_sentenciais = novas_formas
        formas_sentenciais.extend(novas_formas)
    return set(formas_sentenciais)

# Exemplo de uso
gramatica = {'S': ['0S33', 'A'], 'A': ['1']}
simbolo_inicial = 'S'
max_derivacoes = 2
formas_sentenciais = gerar_derivacoes(gramatica, simbolo_inicial, max_derivacoes)
print("Formas sentenciais geradas:", formas_sentenciais)


Formas sentenciais geradas: {'0A33', '0S33', 'A', '1', 'S', '00S3333'}


#### outra forma de resolver

In [53]:
def gerar_formas_sentenciais(gramatica, simbolo_inicial, max_derivacoes):
    formas_sentenciais = [simbolo_inicial]
    for _ in range(max_derivacoes):
        novas_formas = []
        for forma in formas_sentenciais:
            for simbolo in forma:
                if simbolo in gramatica:
                    for producao in gramatica[simbolo]:
                        novas_formas.append(forma.replace(simbolo, producao, 1))
        formas_sentenciais.extend(novas_formas)
    return set(formas_sentenciais)

# Exemplo de uso
gramatica = {'S': ['0S33', 'A'], 'A': ['1']}
simbolo_inicial = 'S'
max_derivacoes = 3
formas_sentenciais = gerar_formas_sentenciais(gramatica, simbolo_inicial, max_derivacoes)
print("Formas sentenciais geradas:", formas_sentenciais)


Formas sentenciais geradas: {'0133', '00S3333', '0S33', '00A3333', '000S333333', 'A', 'S', '0A33', '1'}


### Continuação

#### Exemplos

Dado $G_1 = (V_1, \Sigma_1, P_1, S)$, com:

- $V_1 = \{ 0,1,2,3,S,A \}$
- $\Sigma_1 = \{ 0, 1, 2, 3 \}$
- $N_1 = \{ S, A \}$
- $P_1 = \{ S \to 0S33, S \to A, A \to 12, A \to \epsilon \}$


Considerando a gramática $G_1$ definida anteriormente, temos os seguintes exemplos de derivações e sentenças:

- Derivações não-triviais:
  - $S \overset{+}{\implies} 0S33$: Aplicando uma vez a produção $S \to 0S33$.
  - $S \overset{+}{\implies} 00A3333$: Aplicando duas vezes a produção $S \to 0S33$, seguido por $S \to A$.
  
- Derivações:
  - $00S3333 \overset{*}{\implies} 00S3333$: Aqui, não aplicamos nenhuma produção, portanto, a cadeia permanece a mesma.
  - $0S33 \overset{*}{\implies} 00A3333$: Aplicando primeiro $S \to 0S33$ e depois $S \to A$.

- Sentenças:
  - $12$: Obtida diretamente pela produção $A \to 12$.
  - $00123333$: Obtida por uma série de derivações a partir de $S$, passando por $0S33$, $00S3333$, $00A3333$ e finalmente $00123333$.





### Linguagem definida pela Gramática G

Quando falamos sobre a "linguagem definida por uma gramática $G$", estamos nos referindo ao conjunto completo de sentenças que podem ser geradas usando as regras dessa gramática. Essas sentenças são sequências de símbolos terminais, que são os elementos finais da linguagem.

**Formalmente**

A linguagem $L(G)$ de uma gramática $G$ é o conjunto de todas as sentenças $w$ que podem ser formadas a partir do símbolo inicial $S$ da gramática, utilizando suas regras de produção:

$L(G) = \{ w \in \Sigma^* | S \overset{+}{\implies} w \}$

- $\Sigma^*$ representa o conjunto de todas as possíveis sequências de símbolos terminais.
- $S \overset{+}{\implies} w$ significa que a sentença $w$ pode ser derivada do símbolo inicial $S$ através de uma ou mais aplicações das regras de produção.


#### Exemplo com $G_1$


Para a gramática $G_1$, a linguagem $L_1(G_1)$ pode ser expressa como:

$L_1(G_1) = \{ 0^m 1^n 2^n 3^{2m} | m \geq 0\ e\ (n=0\ ou\ n=1 )\}$

Isso significa que as sentenças em $L_1(G_1)$ são formadas por uma série de '0's seguidos por um número igual de '1's e '2's e, finalmente, o dobro do número de '0's em '3's. Aqui, $m$ pode ser zero ou mais, enquanto $n$ pode ser zero ou um.

**Exemplos de Sentenças**

Exemplos de sentenças que podem ser geradas por $G_1$ incluem:

- $\epsilon$ (cadeia vazia, para $m = 0$ e $n = 0$)
- $12$ (para $m = 0$ e $n = 1$)
- $033$ (para $m = 1$ e $n = 0$)
- $01233$ (para $m = 1$ e $n = 1$)
- $003333$ (para $m = 2$ e $n = 0$)
- $00123333$ (para $m = 2$ e $n = 1$)

Estes exemplos ilustram como as regras de produção de $G_1$ são aplicadas para criar sentenças válidas na linguagem definida pela gramática.

#### Gramática $G_2$

Vamos analisar a gramática $G_2$ e entender como a sentença "aabbcc" pode ser gerada a partir dela.

**Definição de $G_2$**

Considere a gramática $G_2 = (V_2, \Sigma_2, P_2, S)$, definida como:

- $V_2 = \{ a, b, c, S, B, C \}$
  - Este é o conjunto de todos os símbolos utilizados na gramática, incluindo tanto símbolos terminais quanto não terminais.
- $\Sigma_2 = \{ a, b, c \}$
  - Este é o conjunto de símbolos terminais, ou seja, os elementos que aparecem na forma final das sentenças geradas pela gramática.
- $P_2$ contém as regras de produção:
  - $S \to aSBC$
  - $S \to abC$
  - $CB \to BC$
  - $bB \to bb$
  - $bC \to bc$
  - $cC \to cc$
  - Estas regras definem como os símbolos não terminais podem ser substituídos para formar sentenças.

**Linguagem $L_2(G_2)$**

A linguagem gerada por $G_2$ é denotada por $L_2(G_2)$ e consiste em todas as sentenças que podem ser geradas seguindo as regras de produção de $G_2$:

- $L_2(G_2) = \{ a^n b^n c^n | n \geq 1 \}$
  - Isso significa que a linguagem contém sentenças com números iguais de 'a's, 'b's e 'c's em sequência, começando com pelo menos um de cada.

**Geração da Sentença "aabbcc"**

Para entender como a sentença "aabbcc" é gerada usando $G_2$, siga estas etapas de derivação:

1. **Comece com o símbolo inicial $S$.**
   - Iniciamos com o símbolo inicial da gramática, $S$.

2. **Aplicar a regra $S \to aSBC$**
   - Substituindo $S$ por "aSBC", obtemos "`aSBC`".

3. **Aplicar a regra $S \to abC$**
   - Na cadeia "aSBC", substituímos o $S$ por "abC", resultando em "a`abC`BC".

4. **Aplicar a regra $CB \to BC$**
   - Na cadeia "aabCBC", aplicamos a regra $CB \to BC$ para trocar a posição dos símbolos, levando a "aab`BC`C".

5. **Aplicar a regra $bB \to bb$**
   - Usando a regra $bB \to bb$, transformamos "aabBCC" em "aa`bb`CC".

6. **Aplicar a regra $bC \to bc$**
   - Em seguida, aplicamos a regra $bC \to bc$ em "aabbCC", o que resulta em "aab`bc`C".

7. **Aplicar a regra $cC \to cc$**
   - Finalmente, aplicando $cC \to cc$ em "aabbcC", obtemos "aabb`cc`".

Portanto, a sequência correta de derivações para obter "aabbcc" de $G_2$ é:

$S \implies aSBC \implies aaSBCC \implies aabCBCC \implies aabBCCC \implies aabbCCC \implies aabbcc$

Isso é realizado pelas produções:

$S \to aSBC, S \to abC, CB \to BC, bB \to bb, cC \to cc$


Com isso, seguimos as etapas corretas de derivação usando as produções fornecidas pela gramática $G_2$, alcançando a sentença "aabbcc".
É importante notar que a formação da sentença "aabbcc" da gramática $G_2$ envolve uma sequência estratégica de aplicações das regras de produção, garantindo que cada 'a', 'b', e 'c' esteja na ordem correta e em números iguais.



#### Exemplo
Podemos fazer um código que faça uma geração direta de todas as cadeias. Mas essa resolução adiciona mais complexidade.


In [2]:
# Definindo o alfabeto
sigma = ['a', 'b']

# Função para gerar e imprimir cadeias do alfabeto até um comprimento máximo
def generate_and_print_chains(alfabeto, max_length):
    chains = ['']  # Inicia com a cadeia vazia (para Sigma^0)
    print("Início: Sigma^0 = {''}")

    for length in range(1, max_length + 1):
        new_chains = []
        print(f"\nGerando Sigma^{length} (comprimento {length}):")
        
        for chain in chains:
            for character in alfabeto:
                new_chain = chain + character
                new_chains.append(new_chain)
                print(f"Adicionando '{new_chain}' à lista de cadeias")
        
        print(f"Sigma^{length} = {new_chains}")
        chains.extend(new_chains)  # Extenda após a impressão para manter o rastreamento correto

    return chains

# Gerando e exibindo cadeias até o comprimento 3
sigma_star = generate_and_print_chains(sigma, 3)
print("\nCadeias de caracteres formadas pelo alfabeto até o comprimento 3 (Sigma^*):")
print(sigma_star)


Início: Sigma^0 = {''}

Gerando Sigma^1 (comprimento 1):
Adicionando 'a' à lista de cadeias
Adicionando 'b' à lista de cadeias
Sigma^1 = ['a', 'b']

Gerando Sigma^2 (comprimento 2):
Adicionando 'a' à lista de cadeias
Adicionando 'b' à lista de cadeias
Adicionando 'aa' à lista de cadeias
Adicionando 'ab' à lista de cadeias
Adicionando 'ba' à lista de cadeias
Adicionando 'bb' à lista de cadeias
Sigma^2 = ['a', 'b', 'aa', 'ab', 'ba', 'bb']

Gerando Sigma^3 (comprimento 3):
Adicionando 'a' à lista de cadeias
Adicionando 'b' à lista de cadeias
Adicionando 'aa' à lista de cadeias
Adicionando 'ab' à lista de cadeias
Adicionando 'ba' à lista de cadeias
Adicionando 'bb' à lista de cadeias
Adicionando 'aa' à lista de cadeias
Adicionando 'ab' à lista de cadeias
Adicionando 'ba' à lista de cadeias
Adicionando 'bb' à lista de cadeias
Adicionando 'aaa' à lista de cadeias
Adicionando 'aab' à lista de cadeias
Adicionando 'aba' à lista de cadeias
Adicionando 'abb' à lista de cadeias
Adicionando 'baa' à

#### Questão 5

Parece que há uma confusão está no loop que gera as novas cadeias, pois cadeias de passos anteriores estão sendo  repetidas. Dessa forma, precisamos garantir que cada conjunto de cadeias de um determinado comprimento seja gerado corretamente antes de prosseguir para o próximo comprimento. Corrija o problema com uma linha de código

In [5]:
# Definindo o alfabeto
sigma = ['a', 'b']

# Função para gerar e imprimir cadeias do alfabeto até um comprimento máximo
def generate_and_print_chains(alfabeto, max_length):
    chains = ['']  # Inicia com a cadeia vazia (para Sigma^0)
    print("Início: Sigma^0 = {''}")

    for length in range(1, max_length + 1):
        new_chains = []
        print(f"\nGerando Sigma^{length} (comprimento {length}):")
        # 2**2 = 4
        # Gere cadeias de caracteres apenas a partir das cadeias do comprimento anterior
        previous_chains = chains[-(len(alfabeto)**(length-1)):] if length > 1 else ['']
        
        for chain in previous_chains:
            for character in alfabeto:
                new_chain = chain + character
                new_chains.append(new_chain)
                print(f"Adicionando '{new_chain}' à lista de cadeias")
        
        chains.extend(new_chains)
        print(f"Sigma^{length} = {new_chains}")

    return chains

# Gerando e exibindo cadeias até o comprimento 3
sigma_star = generate_and_print_chains(sigma, 3)
print("\nCadeias de caracteres formadas pelo alfabeto até o comprimento 3 (Sigma^*):")
print(sigma_star)


Início: Sigma^0 = {''}

Gerando Sigma^1 (comprimento 1):
Adicionando 'a' à lista de cadeias
Adicionando 'b' à lista de cadeias
Sigma^1 = ['a', 'b']

Gerando Sigma^2 (comprimento 2):
Adicionando 'aa' à lista de cadeias
Adicionando 'ab' à lista de cadeias
Adicionando 'ba' à lista de cadeias
Adicionando 'bb' à lista de cadeias
Sigma^2 = ['aa', 'ab', 'ba', 'bb']

Gerando Sigma^3 (comprimento 3):
Adicionando 'aaa' à lista de cadeias
Adicionando 'aab' à lista de cadeias
Adicionando 'aba' à lista de cadeias
Adicionando 'abb' à lista de cadeias
Adicionando 'baa' à lista de cadeias
Adicionando 'bab' à lista de cadeias
Adicionando 'bba' à lista de cadeias
Adicionando 'bbb' à lista de cadeias
Sigma^3 = ['aaa', 'aab', 'aba', 'abb', 'baa', 'bab', 'bba', 'bbb']

Cadeias de caracteres formadas pelo alfabeto até o comprimento 3 (Sigma^*):
['', 'a', 'b', 'aa', 'ab', 'ba', 'bb', 'aaa', 'aab', 'aba', 'abb', 'baa', 'bab', 'bba', 'bbb']


# Slide 50 - Gramáticas Equivalentes

## Exemplos 1

**Gramáticas Equivalentes**

Duas ou mais gramáticas são consideradas **equivalentes** quando definem a mesma linguagem. Isso significa que, apesar de possivelmente terem regras de produção diferentes, elas geram o mesmo conjunto de sentenças.

**Exemplo de Gramáticas Equivalentes**

Considere as gramáticas $G_3$ e $G_4$:

- $G_3 = (\{a, b, S\}, \{a, b\}, \{S \to aS, S \to a, S \to bS, S \to b, S \to aSb\}, S)$
- $G_4 = (\{a, b, S, X \}, \{a, b\}, \{S \to XS, S \to X, X \to a, X \to b\}, S)$

**Linguagem Definida por $G_3$ e $G_4$**

Ambas as gramáticas $G_3$ e $G_4$ definem a mesma linguagem, que pode ser expressa como $\{a, b\}^+$. Isso significa que a linguagem consiste em uma ou mais combinações dos símbolos 'a' e 'b'.

- Em $G_3$, as regras permitem a geração de cadeias compostas por 'a' e 'b', onde 'a' e 'b' podem aparecer sozinhos ou seguidos por $S$, e até mesmo 'a' seguido por 'b' através de $S$.
- Em $G_4$, a introdução de $X$ como um passo intermediário serve para gerar 'a' ou 'b', e $S$ pode ser seguido por mais $XS$, permitindo uma construção similar à de $G_3$.

**Conclusão:**

As gramáticas $G_3$ e $G_4$ são equivalentes porque ambas definem a mesma linguagem de cadeias não vazias de 'a' e 'b', indicadas por $\{a, b\}^+$, onde o símbolo '$^+$' denota uma ou mais repetições dos elementos 'a' e 'b'.


**Gramáticas Equivalentes $G_3$ e $G_4$**

Duas gramáticas são consideradas equivalentes se elas geram a mesma linguagem, ou seja, o conjunto de todas as sentenças que podem ser derivadas de cada gramática é o mesmo.

**Gramáticas Definidas**

- $G_3$ é definida como:
  - Conjunto de símbolos: $\{a, b, S\}$
  - Símbolos terminais: $\{a, b\}$
  - Regras de produção:
    - $S \to aS$
    - $S \to a$
    - $S \to bS$
    - $S \to b$
    - $S \to aSb$
  - Símbolo inicial: $S$

- $G_4$ é definida como:
  - Conjunto de símbolos: $\{a, b, S, X\}$
  - Símbolos terminais: $\{a, b\}$
  - Regras de produção:
    - $S \to XS$
    - $S \to X$
    - $X \to a$
    - $X \to b$
  - Símbolo inicial: $S$

**Linguagem Gerada por $G_3$ e $G_4$**

Ambas as gramáticas geram a linguagem $\{a, b\}^+$, que representa todas as cadeias possíveis compostas de 'a' e 'b', com pelo menos um símbolo presente.

**Exemplos de Derivações**

**Em $G_3$**

Para gerar a cadeia "aab", podemos seguir estas etapas de derivação em $G_3$:

1. $S \to aS$
2. $aS \to aaS$
3. $aaS \to aab$

**Em $G_4$**

Para gerar a mesma cadeia "aab", as etapas de derivação em $G_4$ seriam:

1. $S \to XS$
2. $XS \to aS$
3. $aS \to aXS$
4. $aXS \to aaS$
5. $aaS \to aab$

**Conclusão**

Embora as regras de produção em $G_3$ e $G_4$ sejam diferentes, ambas as gramáticas permitem gerar todas as possíveis combinações de 'a' e 'b', demonstrando que elas são equivalentes ao definir a mesma linguagem $\{a, b\}^+$.


## Exemplos 2

**Mais exemplos de Gramáticas e suas Linguagens**

Considere as seguintes gramáticas e as linguagens que definem:

- Gramática $G_0$
    - **Gramática**: $G_0 = (\{a, b, S\}, \{a, b\}, \{S \to aS, S \to bS, S \to \epsilon\}, S)$
    - **Linguagem**: $L_0(G_0) = \Sigma^*$
    - Essa gramática gera todas as cadeias possíveis, incluindo a cadeia vazia, sobre o alfabeto $\{a, b\}$.

- Gramática $G_1$
    - **Gramática**: $G_1 = (\{a, b, S\}, \{a, b\}, \{S \to aS, S \to bS, S \to a, S \to b\}, S)$
    - **Linguagem**: $L_1(G_1) = \Sigma^+$
    - Esta gramática gera todas as cadeias não vazias sobre o alfabeto $\{a, b\}$.

- Gramática $G_2$
    - **Gramática**: $G_2 = (\{a, b, S, X \}, \{a, b\}, \{S \to aX, X \to aX, X \to bX, X \to \epsilon\}, S)$
    - **Linguagem**: $L_2(G_2)$ representa todas as cadeias sobre $\{a,b\}$ que começam com $a$.

- Gramática $G_3$
    - **Gramática**: $G_3 = (\{a, b, S, X \}, \{a, b\}, \{S \to aX, X \to aX, X \to bX, X \to b\}, S)$
    - **Linguagem**: $L_3(G_3)$ define todas as cadeias sobre $\{a,b\}$ que começam com $a$ e terminam com $b$.

- Gramática $G_4$
    - **Gramática**: $G_4 = (\{a, b, S, X \}, \{a, b\}, \{S \to XbXbX, X \to aX, X \to \epsilon\}, S)$
    - **Linguagem**: $L_4(G_4)$ engloba todas as cadeias sobre $\{a,b\}$ que contêm exatamente dois $bs$.

- Gramática $G_5$
    - **Gramática**: $G_5 = (\{a, b, S, X \}, \{a, b\}, \{S \to bX, X \to aX, X \to \epsilon\}, S)$
    - **Linguagem**: $L_5(G_5)$ inclui todas as cadeias sobre $\{a,b\}$ que começam com um único $b$ e tem 0 ou mais $as$.

Cada uma dessas gramáticas especifica regras que formam as estruturas das cadeias em suas respectivas linguagens, ilustrando a diversidade e a flexibilidade das gramáticas na definição de conjuntos específicos de cadeias.


**Detalhamento das Gramáticas e Suas Linguagens Equivalentes**

**Gramática $G_0$**
- **Gramática**: $G_0 = (\{a, b, S\}, \{a, b\}, \{S \to aS, S \to bS, S \to \epsilon\}, S)$
- **Linguagem**: $L_0(G_0) = \Sigma^*$
  - Esta gramática gera todas as cadeias possíveis, incluindo a cadeia vazia, sobre o alfabeto $\{a, b\}$.
  - **Exemplo de derivação**:
    - $S \to aS \to abS \to ab$

**Gramática $G_1$**
- **Gramática**: $G_1 = (\{a, b, S\}, \{a, b\}, \{S \to aS, S \to bS, S \to a, S \to b\}, S)$
- **Linguagem**: $L_1(G_1) = \Sigma^+$
  - Esta gramática gera todas as cadeias não vazias sobre o alfabeto $\{a, b\}$.
  - **Exemplo de derivação**:
    - $S \to aS \to aaS \to aab$

**Gramática $G_2$**
- **Gramática**: $G_2 = (\{a, b, S, X \}, \{a, b\}, \{S \to aX, X \to aX, X \to bX, X \to \epsilon\}, S)$
- **Linguagem**: $L_2(G_2)$ representa todas as cadeias sobre $\{a,b\}$ que começam com $a$.
  - **Exemplo de derivação**:
    - $S \to aX \to aaX \to aab$

**Gramática $G_3$**
- **Gramática**: $G_3 = (\{a, b, S, X \}, \{a, b\}, \{S \to aX, X \to aX, X \to bX, X \to b\}, S)$
- **Linguagem**: $L_3(G_3)$ define todas as cadeias sobre $\{a,b\}$ que começam com $a$ e terminam com $b$.
  - **Exemplo de derivação**:
    - $S \to aX \to aaX \to aab$

**Gramática $G_4$**
- **Gramática**: $G_4 = (\{a, b, S, X \}, \{a, b\}, \{S \to XbXbX, X \to aX, X \to \epsilon\}, S)$
- **Linguagem**: $L_4(G_4)$ engloba todas as cadeias sobre $\{a,b\}$ que contêm exatamente dois $b$s.
  - **Exemplo de derivação**:
    - $S \to XbXbX \to abXbX \to abbX \to abb$

**Gramática $G_5$**
- **Gramática**: $G_5 = (\{a, b, S, X \}, \{a, b\}, \{S \to bX, X \to aX, X \to \epsilon\}, S)$
- **Linguagem**: $L_5(G_5)$ inclui todas as cadeias sobre $\{a,b\}$ que começam com um único $b$.
  - **Exemplo de derivação**:
    - $S \to bX \to baX \to baa$

**Conclusão sobre Equivalências**
- $G_0$ e $G_1$ são semelhantes, mas $G_0$ inclui a cadeia vazia (representada por $\epsilon$), enquanto $G_1$ não.
- As gramáticas $G_2$, $G_3$, $G_4$ e $G_5$ definem subconjuntos específicos do alfabeto $\{a, b\}$, com restrições como cadeias que começam ou terminam com um determinado símbolo ou contêm um número exato de um símbolo.


# Old

#### Questão 6

Neste código, ajustamos a lógica para que previous_chains armazene as cadeias do comprimento imediatamente anterior, evitando a repetição de cadeias ao gerar as de comprimento superior. Agora, para cada novo comprimento, o programa concatena caracteres apenas às cadeias do comprimento anterior, o que deve eliminar as repetições e erros na geração das cadeias. Podemos ainda usar uma estrutura de dados chamada `set`. Essa estrutura de dados evita de guardarmos elementos repetidos, mesmo que tentemos fazer. Resolva o problema usando essa estrutura de dados

In [6]:
# Definindo o alfabeto
sigma = ['a', 'b']

# Função para gerar e imprimir cadeias do alfabeto até um comprimento máximo
def generate_and_print_chains(alfabeto, max_length):
    chains = set([''])  # Inicia com a cadeia vazia (para Sigma^0) em um conjunto
    print("Início: Sigma^0 = {''}")

    for length in range(1, max_length + 1):
        print(f"\nGerando Sigma^{length} (comprimento {length}):")
        
        new_chains = set()
        for chain in chains:
            for character in alfabeto:
                new_chain = chain + character
                new_chains.add(new_chain)  # Adiciona a nova cadeia ao conjunto
                print(f"Adicionando '{new_chain}' à lista de cadeias")

        chains = chains.union(new_chains)  # Une o conjunto atual com o novo conjunto de cadeias
        print(f"Sigma^{length} = {list(new_chains)}")

    return chains

# Gerando e exibindo cadeias até o comprimento 3
sigma_star = generate_and_print_chains(sigma, 3)
print("\nCadeias de caracteres formadas pelo alfabeto até o comprimento 3 (Sigma^*):")
print(sorted(sigma_star))  # Convertendo o conjunto em uma lista ordenada para impressão


Início: Sigma^0 = {''}

Gerando Sigma^1 (comprimento 1):
Adicionando 'a' à lista de cadeias
Adicionando 'b' à lista de cadeias
Sigma^1 = ['a', 'b']

Gerando Sigma^2 (comprimento 2):
Adicionando 'a' à lista de cadeias
Adicionando 'b' à lista de cadeias
Adicionando 'aa' à lista de cadeias
Adicionando 'ab' à lista de cadeias
Adicionando 'ba' à lista de cadeias
Adicionando 'bb' à lista de cadeias
Sigma^2 = ['bb', 'ba', 'aa', 'b', 'ab', 'a']

Gerando Sigma^3 (comprimento 3):
Adicionando 'a' à lista de cadeias
Adicionando 'b' à lista de cadeias
Adicionando 'bba' à lista de cadeias
Adicionando 'bbb' à lista de cadeias
Adicionando 'baa' à lista de cadeias
Adicionando 'bab' à lista de cadeias
Adicionando 'aaa' à lista de cadeias
Adicionando 'aab' à lista de cadeias
Adicionando 'ba' à lista de cadeias
Adicionando 'bb' à lista de cadeias
Adicionando 'aba' à lista de cadeias
Adicionando 'abb' à lista de cadeias
Adicionando 'aa' à lista de cadeias
Adicionando 'ab' à lista de cadeias
Sigma^3 = ['baa

## Slide 18 - Operações e Propriedades

### Questão 7  
Considere os conjuntos de strings $L = \{ "001", "10", "111" \}$ e $M = \{ \epsilon, "001" \}$, onde $\epsilon$ representa a string vazia. Escreva um programa em Python que calcule a concatenação de $L$ e $M$, representada por $LM$. $LM$ é o conjunto de todas as strings formadas pela concatenação de cada elemento em $L$ com cada elemento em $M$.

#### Resolução 1  
Usando `set`

In [13]:
# Definindo os conjuntos de strings
L = {"001", "10", "111"}
M = {"", "001"}

# Função para concatenar dois conjuntos de strings
def concatenate_sets(set1, set2):
    concatenated_set = set()
    for element1 in set1:
        for element2 in set2:
            concatenated_set.add(element1 + element2)
    return concatenated_set

# Calculando a concatenação de L e M
LM = concatenate_sets(L, M)

# Exibindo o resultado
print("LM =", LM)


LM = {'10001', '001', '10', '111', '001001', '111001'}


**Explicação**
- `L` e `M` são definidos como conjuntos de strings.
- A função `concatenate_sets` recebe dois conjuntos, `set1` e `set2`, e itera sobre cada elemento de set1 e set2 para concatenar os elementos correspondentes, formando assim um novo conjunto que contém todos os resultados da concatenação.
- A concatenação de `L` e `M (LM)` é então calculada usando essa função.
- O resultado, `LM`, é impresso e deve conter todas as strings resultantes da concatenação de cada string em `L` com cada string em `M`.
- No exemplo dado, `LM` incluirá as strings `"001", "001001", "10", "10001", "111", e "111001"`, correspondendo à concatenação de todos os elementos de `L` com todos os elementos de `M`.

#### Resolução 2
Usando `itertools.product` para realizar o produto cartesiano

In [14]:
# Definindo os conjuntos de strings
L = {"001", "10", "111"}
M = {"", "001"}

# Usando produto cartesiano para concatenar os conjuntos
import itertools

LM = {''.join(pair) for pair in itertools.product(L, M)}

# Exibindo o resultado
print("LM =", LM)

LM = {'10001', '001', '10', '111', '001001', '111001'}


- **`L` e `M`** são definidos como conjuntos de strings, com `M` contendo a string vazia $\epsilon$ e "001".
- Utilizamos a função `product` do módulo `itertools` para criar o produto cartesiano de `L` e `M`. Isso nos dá todos os pares possíveis de elementos entre `L` e `M`.
- A compreensão de conjunto `{''.join(pair) for pair in itertools.product(L, M)}` concatena cada par de strings formando o conjunto `LM`.
- `LM` contém todas as strings resultantes da concatenação de elementos de `L` com elementos de `M`, como "001", "001001", "10", "10001", "111", e "111001".


## Slide 19 - Fechamento reflexivo e transitivo

### Linguagem

O **fechamento reflexivo e transitivo** é um conceito da teoria de linguagens formais e autômatos, geralmente aplicado a conjuntos de cadeias ou relações. Ou seja, dada uma linguagem $L$, é basicamente uma maneira de obter todas as possíveis cadeias (sequências de símbolos) que você pode formar usando as cadeias originais em $L$, quantas vezes quiser, incluindo a possibilidade de não usar nenhuma cadeia (que é representada pela cadeia vazia $ϵ$). Vamos explorar esse conceito em detalhes e compará-lo com a reflexividade e transitividade em funções ou relações.

**Fechamento Reflexivo**
- **Linguagens Formais**: No contexto de linguagens formais, o fechamento reflexivo de uma linguagem $L$ permite a inclusão da cadeia vazia $\epsilon$. Isso significa que, independentemente de não selecionar nenhum elemento de $L$, a cadeia vazia é considerada uma concatenação válida de zero elementos de $L$.
- **Relações**: Em teoria de relações, uma relação é considerada reflexiva se cada elemento estiver relacionado a si mesmo. Por exemplo, na relação de igualdade, cada número é igual a ele mesmo.

**Fechamento Transitivo**
- **Linguagens Formais**: O fechamento transitivo de uma linguagem $L$ envolve criar novas cadeias por meio da concatenação repetida de elementos de $L$. Se $L = \{a, b\}$, então $L^*$ (o fechamento transitivo de $L$) incluiria cadeias como $aa$, $ab$, $ba$, $bb$, $aaa$, e assim por diante, abrangendo todas as concatenações possíveis dos elementos de $L$.
- **Relações**: Na teoria de relações, transitividade implica que se um elemento $a$ está relacionado a um elemento $b$, e $b$ está relacionado a um elemento $c$, então $a$ deve estar relacionado a $c$. Isso é diferente da concatenação em linguagens, que se concentra na combinação de cadeias.

**Diferenças e Semelhanças**
- **Operação vs. Propriedade**: O fechamento reflexivo e transitivo em linguagens é uma operação que gera um conjunto novo a partir do original, enquanto reflexividade e transitividade em relações são propriedades que descrevem como os elementos de um conjunto estão interconectados.
- **Resultado vs. Condição**: No fechamento reflexivo e transitivo, estamos interessados no resultado (o conjunto de todas as cadeias possíveis), enquanto em reflexividade e transitividade, o foco está em se a relação satisfaz certas condições.
- **Completude**: Ambos os conceitos lidam com a ideia de completude. No fechamento reflexivo e transitivo, trata-se da completude das cadeias que podem ser formadas; nas relações reflexivas e transitivas, é sobre a completude das conexões dentro do conjunto.

Essas explicações ilustram como o fechamento reflexivo e transitivo expande um conjunto de cadeias em linguagens formais, enquanto a reflexividade e a transitividade em relações descrevem as conexões dentro de um conjunto.

Se você tem um conjunto $L$ com algumas cadeias de caracteres, então $L∗$ incluirá:
- A cadeia vazia $ϵ$
- Todas as cadeias em $L$ (isso é $L^1$)
- Todas as cadeias que podem ser formadas concatenando duas cadeias em $L$ (isso é $L^2$)
E assim por diante, para 3 cadeias, 4 cadeias, etc.


#### Questão 4  
Seja $L$ o conjunto de todas as cadeias formadas apenas pelo símbolo '0'. Calcule $L^*$, o fechamento reflexivo e transitivo de $L$, o que representa todas as cadeias que podem ser formadas concatenando zero ou mais vezes as cadeias em $L$.

In [6]:
# Definindo a linguagem L
L = {'aba', 'bab'}	

# Função para calcular o fechamento reflexivo e transitivo de L
def reflexive_transitive_closure(language, max_length=5):
    closure = {''}  # Inclui a cadeia vazia para o fechamento reflexivo
    print("Início: L^0 = {''}")
    
    for i in range(1, max_length + 1):
        new_elements = set()
        for element in closure:
            for char in language:
                new_chain = element + char
                if new_chain not in closure:
                    new_elements.add(new_chain)
                    print(f"Adicionando '{new_chain}' à L^{i}")
        
        if not new_elements:
            print(f"Nenhum novo elemento adicionado para L^{i}")
            break
        
        closure.update(new_elements)
        print(f"L^{i} = {sorted(closure)}")

    return closure

# Calculando L*
L_star = reflexive_transitive_closure(L)
print("\nL* =", sorted(L_star))


Início: L^0 = {''}
Adicionando 'bab' à L^1
Adicionando 'aba' à L^1
L^1 = ['', 'aba', 'bab']
Adicionando 'ababab' à L^2
Adicionando 'abaaba' à L^2
Adicionando 'babbab' à L^2
Adicionando 'bababa' à L^2
L^2 = ['', 'aba', 'abaaba', 'ababab', 'bab', 'bababa', 'babbab']
Adicionando 'babbabbab' à L^3
Adicionando 'babbababa' à L^3
Adicionando 'abaababab' à L^3
Adicionando 'abaabaaba' à L^3
Adicionando 'babababab' à L^3
Adicionando 'bababaaba' à L^3
Adicionando 'abababbab' à L^3
Adicionando 'ababababa' à L^3
L^3 = ['', 'aba', 'abaaba', 'abaabaaba', 'abaababab', 'ababab', 'ababababa', 'abababbab', 'bab', 'bababa', 'bababaaba', 'babababab', 'babbab', 'babbababa', 'babbabbab']
Adicionando 'abaabaababab' à L^4
Adicionando 'abaabaabaaba' à L^4
Adicionando 'abaabababbab' à L^4
Adicionando 'abaababababa' à L^4
Adicionando 'babbabababab' à L^4
Adicionando 'babbababaaba' à L^4
Adicionando 'bababaababab' à L^4
Adicionando 'bababaabaaba' à L^4
Adicionando 'bababababbab' à L^4
Adicionando 'babababababa' à 

### Alfabeto

- **Conjunto Vazio $\varnothing$**: 
  - O conjunto vazio $\varnothing$ contém zero cadeias e representa a menor linguagem possível definida sobre um alfabeto $\Sigma$. Ele é crucial em teoria da computação porque serve como a base para construir linguagens mais complexas, representando a ideia de "nada" ou "ausência" em termos de cadeias de caracteres.

- **Fechamento Reflexivo e Transitivo $\Sigma^*$**:
  - $\Sigma^*$ é o conjunto de todas as possíveis cadeias que podem ser formadas com os elementos de $\Sigma$, incluindo a cadeia vazia $\epsilon$. Este conjunto é a maior de todas as linguagens que se pode definir sobre $\Sigma$, pois inclui todas as combinações possíveis de elementos em $\Sigma$ em qualquer comprimento, começando do zero (cadeia vazia).

- **Conjunto Potência $2^{\Sigma^*}$**:
  - $2^{\Sigma^*}$ representa o conjunto de todos os subconjuntos possíveis formados a partir de $\Sigma^*$, incluindo o próprio $\Sigma^*$ e o conjunto vazio $\varnothing$. Ele corresponde ao conjunto de todas as possíveis linguagens que podem ser definidas sobre $\Sigma$, abrangendo todas as variações e combinações de cadeias possíveis.
  - É interessante notar que $\varnothing$ e $\Sigma^*$ são elementos de $2^{\Sigma^*}$, evidenciando a abrangência deste conjunto, que vai da menor à maior linguagem possível sobre o alfabeto $\Sigma$.

Essas observações realçam a estrutura hierárquica e inclusiva dos conjuntos em teoria das linguagens formais, desde o conjunto mais simples $\varnothing$ até o conjunto potência $2^{\Sigma^*}$, que encapsula todas as linguagens possíveis.

Considere o alfabeto $\Sigma = \{a, b, c\}$ e uma propriedade $P$ que define que todas as cadeias devem iniciar com o símbolo 'a'. Vamos explorar algumas linguagens derivadas de $\Sigma$ e como elas se relacionam com a propriedade $P$:

- **Linguagem $L_0$**:
  - Definida como $L_0 = \varnothing$, é a menor linguagem possível sobre $\Sigma$. Ela não contém nenhuma cadeia e, portanto, não contribui para a propriedade $P$.

- **Linguagem $L_1$**:
  - Contém cadeias que começam com 'a' e segue a propriedade $P$: $L_1 = \{a, ab, ac, abc, acb\}$. Esta é uma linguagem finita e satisfaz a condição de iniciar todas as cadeias com 'a'.

- **Linguagem $L_2$**:
  - É uma linguagem infinita que também obedece a $P$, definida como $L_2 = \{a\}\{a\}^*\{b\}^*\{c\}^*$. Isso significa que $L_2$ começa com 'a', seguido por zero ou mais 'a's, 'b's e 'c's, em qualquer quantidade.

- **Linguagem $L_3$**:
  - A maior linguagem que observa $P$, definida por $L_3 = \{a\}\{a,b,c\}^*$, consiste em todas as cadeias que começam com 'a' seguido por qualquer combinação de 'a', 'b', e 'c'. Ela é infinita e engloba todas as cadeias que podem ser formadas segundo $P$.

- **Subconjuntos e Pertinência**:
  - Todas estas linguagens, $L_0$, $L_1$, $L_2$, e $L_3$, são subconjuntos de $\Sigma^*$ e elementos do conjunto potência $2^{\Sigma^*}$, que representa todas as possíveis linguagens definidas sobre $\Sigma$.

- **Diversidade de Linguagens**:
  - Além de $L_0$, $L_1$, $L_2$ e $L_3$, existem muitas outras linguagens que podem ser construídas a partir de $\Sigma$ seguindo diferentes regras ou propriedades.

Este exemplo demonstra a variedade e a complexidade das linguagens que podem ser definidas a partir de um alfabeto básico, dependendo das propriedades ou regras especificadas, desde o conjunto vazio até linguagens infinitamente expansíveis.

### Questão 5 - Slide 26

### Exemplo de Fechamento Transitivo

Explore o conceito de fechamento transitivo no alfabeto $\Sigma = \{n, (, ), +, *, -, /\}$:

**Resposta**  
O fechamento transitivo é uma operação fundamental em linguagens formais que gera um conjunto contendo todas as possíveis sequências (ou cadeias) de elementos de um alfabeto, concatenadas em qualquer quantidade, incluindo a cadeia vazia ($ϵ$) para o fechamento reflexivo. Vamos explorar esse conceito usando o alfabeto:  

- $\Sigma^*$ representa o conjunto de todas as possíveis cadeias formadas pelos símbolos em $\Sigma$, incluindo a cadeia vazia ($\epsilon$). Exemplos de cadeias em $\Sigma^*$ incluem "n", "n+n", "-n\)", "*/", e "n()\)", representando a vasta gama de expressões que podem ser criadas.

- $\Sigma^+$ é similar ao $\Sigma^*$, mas não inclui a cadeia vazia ($\epsilon$). Assim, contém cadeias como "n", "n+n", "-n)", "n()", e "n-(n*n)", refletindo todas as combinações possíveis de elementos em $\Sigma$ com um ou mais símbolos.

- A relação $\Sigma^+ = \Sigma^* - \{\epsilon\}$ demonstra que $\Sigma^+$ pode ser derivado de $\Sigma^*$ pela remoção da cadeia vazia, ressaltando a conexão entre esses dois conjuntos no contexto do fechamento transitivo.


## Slide 28 - Reversão

### Questão 6

Dada a linguagem $L_2 = \{\epsilon, a, ab, abc\}$, escreva uma função em Python que determine o reverso dessa linguagem, denotado por $L_2^R$. O reverso de uma linguagem é um conjunto de cadeias onde cada cadeia é o inverso das cadeias originais de $L_2$.

In [5]:
# Definindo a linguagem L2
L2 = {'', 'a', 'ab', 'abc'}

# Função para calcular o reverso de uma linguagem
def reverse_language(language):
    return {sentence[::-1] for sentence in language}

# Calculando L2^R
L1 = reverse_language(L2)

# Exibindo o resultado
print("L2^R =", L1)

L2^R = {'', 'ba', 'cba', 'a'}


***Explicação do Código***

- A linguagem $L_2$ é inicialmente definida como um conjunto de cadeias: $\epsilon$ (cadeia vazia), 'a', 'ab', e 'abc'.
- A função `reverse_language` recebe a linguagem $L_2$ como argumento e utiliza a compreensão de conjunto para criar um novo conjunto. Para cada sentença em $L_2$, a sentença é invertida (`sentence[::-1]`) e adicionada ao novo conjunto.
- O resultado `L1` é o conjunto de todas as sentenças de $L_2$ invertidas, ou seja, $L_2^R$.
- Ao imprimir `L1`, obtemos o reverso de $L_2$, que neste caso será $\{\epsilon, a, ba, cba\}$.


## Slide 29 - Propriedade de Prefixo e Sufixo Próprio

**Prefixo Próprio:** Uma cadeia $\alpha$ é um prefixo próprio de outra cadeia $\alpha\beta$ se $\beta$ não é vazia ($\epsilon$). Isso significa que $\alpha$ é o início de $\alpha\beta$, mas $\alpha\beta$ contém mais caracteres além de $\alpha$. Em uma linguagem com a propriedade de prefixo próprio, nenhuma cadeia na linguagem é um prefixo próprio de outra cadeia dentro dessa mesma linguagem.

**Sufixo Próprio:** Similarmente, uma cadeia $\alpha$ é um sufixo próprio de outra cadeia $\beta\alpha$ se $\beta$ não é vazia. Aqui, $\alpha$ aparece no final de $\beta\alpha$, e $\beta\alpha$ contém mais caracteres antes de $\alpha$. Uma linguagem com a propriedade de sufixo próprio não tem nenhuma cadeia que seja sufixo próprio de outra cadeia na linguagem.

Dada a linguagem $L = \{a, ab, abc, bc, c\}$, escreva uma função em Python que determine se $L$ tem a propriedade de prefixo próprio e sufixo próprio. Mostre os elementos que violam estas propriedades, se houver.

In [6]:
L = {'a', 'ab', 'abc', 'bc', 'c'}

# Verificando a propriedade de prefixo próprio
def has_proper_prefix(language):
    for element in language: #a
        for other in language:#ab
            if element != other and other.startswith(element):
                print(f"'{element}' é um prefixo próprio de '{other}'")
                return False
    return True

# Verificando a propriedade de sufixo próprio
def has_proper_suffix(language):
    for element in language:#bc
        for other in language:#abc
            if element != other and other.endswith(element):
                print(f"'{element}' é um sufixo próprio de '{other}'")
                return False
    return True

print("L tem a propriedade de prefixo próprio?", has_proper_prefix(L))
print("L tem a propriedade de sufixo próprio?", has_proper_suffix(L))

'ab' é um prefixo próprio de 'abc'
L tem a propriedade de prefixo próprio? False
'bc' é um sufixo próprio de 'abc'
L tem a propriedade de sufixo próprio? False


### Questão 7

Dada as linguagens abaixo disposta em um dicionário, escreva uma função em Python que determine se cada uma delas tem a propriedade de prefixo próprio e sufixo próprio. Mostre os elementos que violam estas propriedades, se houver.

In [13]:
languages = [
    ('L1', {'prefix', 'pref', 'xfix'}),
    ('L2', {'prefix', 'pref', 'fix'}),
    ('L3', {'alpha', 'beta', 'gamma'}),
    ('L4', {"bat","sat","at"})
]

# Função para verificar a propriedade de prefixo próprio
def has_proper_prefix(language):
    for element in language:
        for other in language:
            if element != other and other.startswith(element):
                print(f"'{element}' é um prefixo próprio de '{other}'")
                return False
    return True

# Função para verificar a propriedade de sufixo próprio
def has_proper_suffix(language):
    for element in language:
        for other in language:
            if element != other and other.endswith(element):
                print(f"'{element}' é um sufixo próprio de '{other}'")
                return False
    return True

# Testando as propriedades para cada linguagem
for name, lang in languages:
    prefix = has_proper_prefix(lang)
    suffix = has_proper_suffix(lang)
    print(f"{name} tem a propriedade de prefixo próprio? {prefix}")
    print(f"{name} tem a propriedade de sufixo próprio? {suffix}")
    if prefix and not suffix:
        print(f"Portanto, {name} tem somente a propriedade de prefixo próprio.\n")
    elif not prefix and suffix:
        print(f"Portanto, {name} tem somente a propriedade de sufixo próprio.\n")
    elif prefix and suffix:
        print(f"Portanto, {name} possui ambas as propriedades de prefixo e sufixo próprio.\n")
    else:
        print(f"Portanto, {name} não possui as propriedades de prefixo e sufixo próprio.\n")


'pref' é um prefixo próprio de 'prefix'
L1 tem a propriedade de prefixo próprio? False
L1 tem a propriedade de sufixo próprio? True
Portanto, L1 tem somente a propriedade de sufixo próprio.

'pref' é um prefixo próprio de 'prefix'
'fix' é um sufixo próprio de 'prefix'
L2 tem a propriedade de prefixo próprio? False
L2 tem a propriedade de sufixo próprio? False
Portanto, L2 não possui as propriedades de prefixo e sufixo próprio.

L3 tem a propriedade de prefixo próprio? True
L3 tem a propriedade de sufixo próprio? True
Portanto, L3 possui ambas as propriedades de prefixo e sufixo próprio.

'at' é um sufixo próprio de 'sat'
L4 tem a propriedade de prefixo próprio? True
L4 tem a propriedade de sufixo próprio? False
Portanto, L4 tem somente a propriedade de prefixo próprio.



## Slide 31 - Quociente de Linguagens

**Definição**
Sejam $L_1$ e $L_2$ duas linguagens. O quociente de $L_1$ por $L_2$, denotado por $L_1 / L_2$, é definido como o conjunto de todas as cadeias $x$ tal que a concatenação de $x$ com qualquer cadeia $y$ de $L_2$ resulta em uma cadeia que pertence a $L_1$.

#### Formulação Matemática
$$
L_1 / L_2 = \{x \mid xy \in L_1, y \in L_2\}
$$

#### Exemplo Prático
Considere as linguagens:
- $L = \{a, aab, baa\}$
- $A = \{a\}$

Para calcular o quociente $L/A$, procuramos por cadeias em $L$ que, ao serem concatenadas com 'a' (os elementos de $A$), ainda pertencem a $L$.

- Da cadeia 'a' em $L$, removendo 'a' de $A$, sobra o $\epsilon$ (cadeia vazia).
- Da cadeia 'aab' em $L$, removendo 'a' de $A$, sobra 'aa'. No entanto, 'aa' não é incluída porque não satisfaz a condição de formar uma cadeia em $L$ ao ser concatenada com 'a'.
- Da cadeia 'baa' em $L$, removendo 'a' de $A$, sobra 'ba'.

Portanto, o quociente $L/A$ é $\{\epsilon, ba\}$.


### Questão 8 - Slide 32

Considere as linguagens seguintes:

- $L_1 = \{a^i b \mid i \geq 0\}$
- $L_2 = \{a^i b c^i \mid i \geq 0\}$
- $L_3 = \{b\}$
- $L_4 = \{a^i b \mid i \geq 1\}$
- $L_5 = \{b c^i \mid i \geq 0\}$
- $L_6 = \{c^i b \mid i \geq 0\}$
- $L_7 = \{a^i \mid i \geq 0\}$

### Descrição das Linguagens

- **$L_1 = \{a^i b \mid i \geq 0\}$**
  
  Esta linguagem consiste em cadeias formadas pela repetição de 'a' zero ou mais vezes seguida por um 'b'. Isso inclui 'b' (quando $i=0$), 'ab' ($i=1$), 'aab' ($i=2$), e assim por diante.

- **$L_2 = \{a^i b c^i \mid i \geq 0\}$**
  
  Aqui, as cadeias começam e terminam com o mesmo número de 'a's e 'c's respectivamente, com um 'b' no meio. Exemplos incluem 'b' ($i=0$), 'abc' ($i=1$), 'aabcc' ($i=2$), etc.

- **$L_3 = \{b\}$**
  
  Esta linguagem é a mais simples e contém uma única cadeia: 'b'.

- **$L_4 = \{a^i b \mid i \geq 1\}$**
  
  Similar a $L_1$, mas aqui temos pelo menos um 'a' antes do 'b', ou seja, não inclui apenas 'b'. Exemplos são 'ab' ($i=1$), 'aab' ($i=2$), etc.

- **$L_5 = \{b c^i \mid i \geq 0\}$**
  
  Consiste em cadeias que começam com 'b' seguidas por zero ou mais 'c's. Inclui 'b' ($i=0$), 'bc' ($i=1$), 'bcc' ($i=2$), e assim por diante.

- **$L_6 = \{c^i b \mid i \geq 0\}$**
  
  Similar a $L_5$, mas aqui 'c's precedem o 'b'. Isso inclui 'b' ($i=0$), 'cb' ($i=1$), 'ccb' ($i=2$), etc.

- **$L_7 = \{a^i \mid i \geq 0\}$**
  
  Esta linguagem inclui somente cadeias de 'a's de qualquer comprimento, incluindo a cadeia vazia $\epsilon$ quando $i=0$, 'a' ($i=1$), 'aa' ($i=2$), e assim por diante.


**Quocientes de Linguagens: Exemplos Detalhados**

1. **$L_1 / L_3 = ?$**

   - $L_1 = \{a^i b \mid i \geq 0\}$ inclui cadeias como 'b', 'ab', 'aab', 'aaab', etc.
   - $L_3 = \{b\}$ contém apenas a cadeia 'b'.

   O quociente $L_1 / L_3$ procura cadeias em $L_1$ que, ao serem concatenadas com 'b' de $L_3$, resultam em cadeias ainda em $L_1$. Portanto, o resultado é composto por cadeias de 'a's de qualquer comprimento.

   **Resposta**: $L_1 / L_3$ são todas as cadeias formadas por 'a', que é $L_7 = \{a^i \mid i \geq 0\}$.

2. **$L_1 / L_4 = ?$**
Considere as seguintes linguagens:

    - $L_1 = \{a^i b \mid i \geq 0\}$: Contém cadeias como 'b', 'ab', 'aab', etc., onde 'b' é precedido por zero ou mais 'a's.
    - $L_4 = \{a^i b \mid i \geq 1\}$: Similar a $L_1$, mas começa com pelo menos um 'a', então não inclui 'b' sozinha. Contém 'ab', 'aab', 'aaab', etc.

        - O Quociente $L_1 / L_4$

            Ao calcular $L_1 / L_4$, procuramos por cadeias em $L_1$ que, ao serem concatenadas com qualquer cadeia de $L_4$, resultem em uma cadeia que ainda pertence a $L_1$.

        -  **Exemplos Detalhados para $L_1 / L_4$**:

            - Cadeia 'b' em $L_1$: Não pode ser formada removendo algo de $L_4$.
            - Cadeia 'ab' em $L_1$: Forma $\epsilon$ ao remover 'ab', então $\epsilon \in L_1 / L_4$.
            - Cadeia 'aab' em $L_1$: Ao remover 'ab', restam 'a', então 'a $\in L_1 / L_4$.
            - Cadeia 'aaab' em $L_1$: Removendo 'ab', restam 'aa', então 'aa' está em $L_1 / L_4$.
            - Cadeia 'aaaab' em $L_1$: Removendo 'ab', restam 'aaa', assim 'aaa' está em $L_1 / L_4$.
            - Continuando assim, todas as cadeias de 'a's de qualquer comprimento estão em $L_1 / L_4$.

    Portanto, $L_1 / L_4$ é composto por todas as cadeias de 'a's, ou seja, é igual a $L_7$.

    **Conclusão:**

    - Cada cadeia em $L_1 / L_4$ é formada pela remoção de 'ab' das cadeias em $L_1$, resultando em cadeias compostas apenas por 'a's.
    - Portanto, **$L_1 / L_4 = L_7 = \{a^i \mid i \geq 0\}$**, que representa todas as possíveis sequências de 'a's, incluindo a cadeia vazia.

3. **$L_5 / L_7 = ?$**

   - $L_5 = \{b c^i \mid i \geq 0\}$ começa com 'b' seguido por zero ou mais 'c's. (b, bc, bcc, bccc)
   - $L_7 = \{a^i \mid i \geq 0\}$ inclui somente cadeias de 'a's de qualquer comprimento ($\epsilon$, a, aa, aaa...)

   Para $L_5 / L_7$, como $L_7$ contém apenas cadeias de 'a's, e nenhuma cadeia em $L_5$ pode resultar em uma cadeia em $L_5$ ao ser concatenada com 'a's, o resultado é o conjunto vazio.

   **Resposta**: $L_5 / L_7$ é $\varnothing$, pois não há como obter cadeias de $L_5$ ao concatenar com cadeias de $L_7$.

4. **$L_2 / L_6 = ?$**

   - $L_2 = \{a^i b c^i \mid i \geq 0\}$ onde o número de 'a's e 'c's é o mesmo, envolvendo um 'b'.  
    (b, abc, aabcc, aaabccc, ...)
   - $L_6 = \{c^i b \mid i \geq 0\}$ contém cadeias que começam com zero ou mais 'c's seguidos por 'b'.  
   (b, cb, ccb, cccb, ...)

   **Processo para encontrar $L_2 / L_6$**

    Ao tentar formar o quociente $L_2 / L_6$, encontramos dificuldades:
    - As cadeias em $L_2$ são estruturadas de forma que 'b' esteja sempre entre o mesmo número de 'a's e 'c's.
    - As cadeias em $L_6$ não se encaixam como sufixos diretos em $L_2$ porque os 'c's em $L_2$ estão sempre precedidos por 'b'.
    - Portanto, não há uma correspondência direta que permita remover um sufixo de $L_6$ de uma cadeia em $L_2$ mantendo a cadeia resultante dentro de $L_2$.

    **Conclusão:** 
    - Nenhuma cadeia em $L_6$ se ajusta perfeitamente como um sufixo removível das cadeias em $L_2$ devido à estrutura específica de $L_2$ que requer um número igual de 'a's e 'c's separados por um único 'b'.
    - Consequentemente, não é possível formar uma cadeia em $L_2$ removendo partes que são consistentes com $L_6$.
    - Dado que não podemos satisfazer a condição para o quociente $L_2 / L_6$ com as cadeias fornecidas em ambas as linguagens, o resultado é o conjunto vazio $\varnothing$.
    - Assim, temos $L_2 / L_6 = \varnothing$, indicando que não existem cadeias em $L_2$ das quais podemos remover um sufixo de $L_6$ e ainda termos uma cadeia que pertença a $L_2$.



## Slide 33 - Substituição

A **substituição** é uma operação fundamental em muitos aspectos da computação e análise de linguagens formais onde você associa cada símbolo de um alfabeto a um conjunto de cadeias de outro alfabeto. Pode-se pensar nisso como uma regra que diz como cada letra de um alfabeto pode ser transformada em palavras de outro alfabeto. Ou seja, é uma maneira de transformar cadeias de um alfabeto em cadeias de outro alfabeto, seguindo um conjunto de regras predefinidas. 

**Definição Formal**
Se temos dois alfabetos:
- $\Sigma_1$: O alfabeto de origem.
- $\Sigma_2$: O alfabeto de destino.

Uma substituição $s$ é uma função que para cada símbolo em $\Sigma_1$, associa um conjunto de cadeias (palavras) formadas pelos símbolos em $\Sigma_2$. Em termos matemáticos, a substituição é descrita como:
$$s: \Sigma_1 \to 2^{\Sigma_2^*}$$
onde $2^{\Sigma_2^*}$ representa o conjunto de todos os subconjuntos de cadeias possíveis em $\Sigma_2$.

**Exemplo Prático**
Considerando:
- $\Sigma_1 = \{a, b, c\}$
- $\Sigma_2 = \{x, y, z\}$

Podemos definir uma substituição $s$ como:
- $s(a) = \{x\}$: O símbolo 'a' é substituído por 'x'.
- $s(b) = \{y, yy\}$: O símbolo 'b' pode ser substituído por 'y' ou 'yy'.
- $s(c) = \{z, zz, zzz\}$: O símbolo 'c' pode ser substituído por 'z', 'zz', ou 'zzz'.

**Como Funciona**
- Quando aplicamos a substituição, cada letra do alfabeto $\Sigma_1$ é transformada nas possíveis cadeias definidas pela função $s$.
- Por exemplo, se temos uma cadeia 'abc' em $\Sigma_1$, após a substituição, poderíamos obter cadeias como 'xyz', 'xyzz', 'xyzzz', etc., dependendo de como escolhemos substituir cada letra.

Agora vamos aplicar essa substituição em várias cadeias do alfabeto $\Sigma_1$:

1. Para a cadeia 'a' em $\Sigma_1$:
   - Após a substituição, obtemos 'x', porque $s(a) = \{x\}$.

2. Para a cadeia 'b' em $\Sigma_1$:
   - Podemos obter 'y' ou 'yy', porque $s(b) = \{y, yy\}$.

3. Para a cadeia 'ab' em $\Sigma_1$:
   - Podemos obter 'xy' ou 'xyy', combinando as possibilidades de substituição para 'a' e 'b'.

4. Para a cadeia 'bc' em $\Sigma_1$:
   - As opções incluem 'yz', 'yzz', 'yzzz', 'yyz', 'yyzz', e 'yyzzz', combinando as substituições para 'b' e 'c'.

5. Para a cadeia 'abc' em $\Sigma_1$:
   - Podemos ter 'xyz', 'xyzz', 'xyzzz', 'xyyz', 'xyyzz', 'xyyzzz', e assim por diante, escolhendo uma substituição para cada símbolo em 'abc'.


Além de aplicar substituições a elementos individuais de um alfabeto, também podemos aplicar uma substituição a uma cadeia inteira. Esta operação é definida de maneira indutiva, ou seja, construída passo a passo, aplicando a substituição a cada símbolo da cadeia.

#### Definição Indutiva
Seja $s$ uma substituição e $w$ uma cadeia, a aplicação de $s$ em $w$, denotada por $s(w)$, segue estas regras:

- Para a cadeia vazia $\epsilon$, $s(\epsilon) = \epsilon$.
- Para uma cadeia $a\alpha$, onde $a$ é um símbolo do alfabeto $\Sigma_1$ e $\alpha$ é uma subsequência de cadeias em $\Sigma_1^*$, temos:
  $$s(a\alpha) = s(a)s(\alpha)$$

Ou seja, a substituição de uma cadeia é o resultado da concatenação das substituições de cada um de seus símbolos.

#### Exemplo Prático

Suponha a cadeia $w = abc$ e a seguinte substituição $s$:

- $s(a) = \{x\}$
- $s(b) = \{y, yy\}$
- $s(c) = \{z, zz, zzz\}$

Então, para aplicar a substituição em $w$:

1. Começamos com $s(abc)$.
2. Aplicamos a substituição ao primeiro símbolo e ao restante da cadeia:
   $$s(abc) = s(a)s(bc)$$
3. Continuamos aplicando a substituição para cada parte:
   $$s(a) = \{x\}$$
   $$s(bc) = s(b)s(c)$$
   $$s(b) = \{y, yy\}$$
   $$s(c) = \{z, zz, zzz\}$$
4. Combinando todas as possíveis substituições de $s(a)$ e $s(b)$ com $s(c)$, obtemos um conjunto de cadeias resultantes de $s(abc)$.

Portanto, $s(abc)$ gera um conjunto de cadeias que inclui combinações como 'xyz', 'xyzz', 'xyzzz', 'xyyz', 'xyyzz', 'xyyzzz', e assim por diante.

Vamos aplicar a substituição em uma cadeia mais complexa, digamos $w = abac$, usando a mesma substituição $s$ do exemplo anterior:

- $s(a) = \{x\}$
- $s(b) = \{y, yy\}$
- $s(c) = \{z, zz, zzz\}$

A cadeia $w = abac$ será analisada e transformada pela substituição $s$.

1. Começamos decompondo a cadeia e aplicando a substituição passo a passo:
   $$s(abac) = s(a)s(bac)$$

2. Aplicando a substituição ao primeiro símbolo e ao restante da cadeia sequencialmente:
   - Para o primeiro 'a': $s(a) = \{x\}$
   - Para o restante 'bac': 
     $$s(bac) = s(b)s(ac)$$

3. Continuamos desdobrando cada parte:
   - Para 'b': $s(b) = \{y, yy\}$
   - Para 'ac':
     $$s(ac) = s(a)s(c)$$
   - Para o segundo 'a': $s(a) = \{x\}$
   - Para 'c': $s(c) = \{z, zz, zzz\}$

4. Agora, combinamos as substituições de cada símbolo para obter o resultado final:
   - Combinando $s(a)$, $s(b)$, e $s(ac)$, temos múltiplas combinações possíveis, dependendo das escolhas em $s(b)$ e $s(c)$.

**Exemplo Concreto de Combinações**
- Considerando uma das possíveis substituições para 'b' e 'c', podemos ter:
  - Se escolhemos 'y' para 'b' e 'z' para 'c', uma das combinações possíveis para $s(abac)$ seria 'xyxz'.
  - Se escolhemos 'yy' para 'b' e 'zzz' para 'c', outra combinação possível seria 'xyyxzzz'.  

Assim, a substituição $s$ é aplicada a cada parte da cadeia $w = abac$, gerando um conjunto de cadeias resultantes que refletem todas as combinações possíveis de substituições conforme definido por $s$.


## Slide 35

A operação de substituição, previamente discutida no contexto de cadeias individuais, pode ser ampliada para ser aplicada a uma linguagem completa. Isso nos permite transformar todas as cadeias em uma linguagem de acordo com regras de substituição específicas.

#### Definição para Linguagens
Seja $s$ uma substituição e $L$ uma linguagem, então a aplicação de $s$ em $L$, denotada por $s(L)$, é definida como:

$$s(L) = \{y | y = s(x)\ para\ x \in L\}$$

Isso significa que para cada cadeia $x$ em $L$, aplicamos a substituição $s$ para gerar uma nova cadeia $y$, e o conjunto de todas essas novas cadeias forma a linguagem $s(L)$.

#### Exemplo Prático

Suponha a linguagem $L = \{a^ib^ic^i | i \geq 1\}$ sobre o alfabeto $\Sigma_1 = \{a,b,c\}$. Cada cadeia em $L$ consiste em 'a's, 'b's e 'c's em número igual, mas variável.

Se definirmos uma substituição $s$ tal que:

- $s(a) = \{x\}$
- $s(b) = \{y, yy\}$
- $s(c) = \{z, zz, zzz\}$

Então, ao aplicar $s$ a $L$, consideramos todas as possíveis combinações geradas pela substituição:

- Para uma cadeia $a^ib^ic^i$ em $L$, onde $i \geq 1$, $s$ mapeia cada 'a' para 'x', cada 'b' para 'y' ou 'yy', e cada 'c' para 'z', 'zz' ou 'zzz'.
- Assim, $s(L)$ incluirá cadeias como $x^iy^jz^k$ onde $i\geq 1$, $i\leq j \leq 2i$, e $i\leq k \leq 3i$ refletindo as múltiplas possibilidades de substituição para 'b' e 'c'.

#### Conclusão

Aplicar uma substituição a uma linguagem permite a transformação de todas as suas cadeias em novas cadeias, seguindo as regras de substituição definidas. Isso resulta em uma nova linguagem que pode ter características e estruturas complexas, dependendo das regras de substituição aplicadas.